# Exercise 4 - A practical case
## I. Abstract
todo
## II. Introduction
<div>
<p></p>
<p style='text-align:justify'>
    Now a day, we can find a high competition between most of the companies from different fields in the business world, this have been increase through the years making companies difficult to reatain the costumers in their services and moreover attract new costumers as a potential clients. Considering the fact that engaging
new customers is often more costly and difficult than retaining existing ones, heads of the companies concerned about this issue are investing in costumer satisfaction polices that can allow to improve the revenue in their business.
<br></br>
Consequently, the strategical and analyst deparment of each company have found that one of the most important factors that assure the steady-state level of customers in a business is the <strong>churn-rate</strong> [1]. This factor is also known as attrition rate and is related to the number of individuals that are quiting their serivices from a determinated company in a certain period. 
</p>
<p style="text-align:justify">In this project we are aiming to determine the churn-rate from a dataset took related to a mobile and communications company <strong>Telecom</strong> in which we have 3333 samples that stand for different costumers in a certain period, and 21 features that are composed by (4 subject attributes, 16 call measurments and their respective churn-rate).
<br></br>
<br></br>
An importan aspect that it has to be considered is that the domain values from each variable/feature quite different in most of the variables, this will lead to the problem due to there will not be and equal contribution for each one of the variables to the data analysis i.e. features with higher values in their scales will influence more in the model when performing the classification task. For that reason it is important to consider a data standarization and prevent this problem. Typical data standardization procedures equalize the range and/or data variability.

The most useful techineques to standarize the data are the following [2]:
<p></p>
<ol>
    <li>Ponderated standarization (0-1 scaling): $$\frac{F_i - min(F_i)}{max(F_i) - min(F_i)}$$</li>
    <li>Dividing each value by the range: $$\frac{F_i}{max(F_i) - min(F_i)}$$</li>
    <li>Normal standarization (z-score scaling): $$\frac{F_i - \mu_i}{\sigma_i}$$</li>
    <li>Dividing each value by the standard deviation: $$\frac{F_i}{\sigma_i}$$</li>
    
</ol>
</p>
<p style="text-align:justify">
Furthermore, we can implemented the data standarization reviewed before considering all the values from the feature or just the values from the feature related to the certain values of other features i.e. considering the normalization of all the values from the total_day_mis, in which state is equal to AZ. This may give an more accurate measurment at data normalization of each feature.
<p></p>
To Continue, theres is another important section of the project related to the metholodogy that is going to be implemented in order to solve the classification problem from the perspective of the supervised learning. The solving methodologies that are suitable for this task are th following:

<ol>
    <li>Adaboost</li>
    <li>Random Forest</li>
    <li>QP solver for SVM considering Stochastic Gradient Descent as an Optimiztion Algorithm </li>
</ol>
</p>

<h3>Adaboost</h3>

<p style="text-align:justify">
A machine learning algorithm based on a supervised approach invented by Yoav Freund a computer science researcher from University of California San Diego in 1995 and redesigned by Robert Shapire a computer science researcher from Princeton University in 1999. The aim of this pattern recognition algorithm is to create a robust classifier based on the linear combination of weak or simple classifiers, as a result of this linear combination there is a non-linear machine learning algorithm, this technique and concepts have been implemented in the well-known "Viola & Jones" that is applied in face recognition application in the computer vision area [3, 4].
<p></p>

<p>Mathematical representation: $$H({x_i}) = {\displaystyle\sum_{j=1}^{T} \alpha_j * h_j({x_i})}$$</p>

<p>Where:</p>
<p>$$
\begin{array}{ll}
H_j({x_i}): classification of the strong classifier\\
x_i: the i^th observation\\ 
T: number of stumps\\ 
\alpha_j: weight of the j^th stumps in linear combination\\
h_j({x_i}): classification of a stump (weak classifier)\\
\end{array} 
$$</p>

<h4>Characteristics: </h4>
<ul>
    <li>A robust non-linear classifier.</li>
    <li>Has a low generalization error.</li>
    <li>Is quite robust to overfitting.</li>
    <li>Is intuitive to implement.</li>
</ul>

<h4>Advantages: </h4>
<ul>
    <li>Find the best or most representative features, can be used as a principal feature selection.</li>
    <li>Find the best thresholds stumps (classifiers).</li>
    <li>Find a strong nad robust classifier.</li>
</ul>

<h4>Considerations: </h4>
<ul>
    <li>Each weak classifier has to have at least 50% of accuracy, which means that requires to have a higher than 0% alpha, in other words it has to have a relevant weight to be considered as part of the strong classifier.</li>
</ul>

<h4>Posible stumps or weak classifiers: </h4>
<ul>
    <li>Decision stump, implementing logistic regression (single axis space partitions).</li>
    <li>Decision tree (hierarchical partitions).</li>
    <li>Random forest (hierarchical partitions).</li>
    <li>Multi-layer perceptron (non-linear approximators).</li>
    <li>Support Vector Machines (linear classifier & Kernell).</li>
</ul>



</p>

<p>
<h3>Random Forests</h3>
<h4>Characteristics: </h4>
<h4>Advantages: </h4>
<h4>Considerations: </h4>
<h4>Posible stumps or weak classifiers: </h4>

</p>

</div>
## III. Specification of the problem and organization of the software
<p>We are the new data analyst in a Telecom company. Our first project is to try to reduce the
churn of our new service. Modeling churn means to understand what keeps the customer engaged
to our product. As an analyst, our goal is to predict or describe the churn rate i.e. the rate at
which customer leave or cease the subscription to a service. Its value lies in the fact that engaging
new customers is often more costly than retaining existing ones. For that reason subscription
business-based companies usually have proactive policies towards customer retention.</p>
<p>In this case study, we aim at building a machine learning model for customer churn prediction
on data from a Telecom company. Each row on the dataset represents a subscribing telephone
customer. Each column contains customer attributes such as phone number, call minutes used
during different times of day, charges incurred for services, lifetime account duration, and whether
or not the customer is still a customer. Data can be found in the file: churn.csv.</p>

The complete set of attributes is the following:
<ul>
<li>State: categorical, for the 50 states and the District of Columbia</li>
<li>Account length: integer-valued, how long an account has been active</li>
<li>Area code: categorical</li>
<li>Phone number: customer ID</li>
<li>International Plan: binary feature, yes or no</li>
<li>VoiceMail Plan: binary feature, yes or no</li>
<li>Number of voice mail messages: integer-valued</li>
<li>Total day minutes: continuous, minutes customer used service during the day</li>
<li>Total day calls: integer-valued</li>
<li>Total day charge: continuous</li>
<li>Total evening minutes: continuous, minutes customer used service during the evening</li>
<li>Total evening calls: integer-valued</li>
<li>Total evening charge: continuous</li>
<li>Total night minutes: continuous, minutes customer used service during the night</li>
<li>Total night calls: integer-valued</li>
<li>Total night charge: continuous</li>
<li>Total international minutes: continuous, minutes customer used service to make international
calls</li>
<li>Total international calls: integer-valued</li>
<li>Total international charge: continuous</li>
<li>Number of calls to customer service: integer-valued</li>
</ul>

## IV. Experiments

### Code Design
All classifiers must be coded from scratch in python, no toolkits are allowed to be used
except for numpy and CVX (http://www.cvxpy.org/). All classifiers have to be fully integrable
in the scikit-learn (sklearn) ecosystem, i.e. inherit from BaseEstimator and implement the basic
methods, fit, predict, score, ... You can check how to integrate your classifier in http://scikit-learn.org/stable/developers/contributing.html#coding-guidelines. We do not need to pass all unit
testing but at least we would like to use several of the nice functions such as gridsearchCV, etc
with your estimator.

<div class = "alert alert-success" style = "border-radius:10px">**WP 1**
<ol>
<li> Read the dataset ‘churn.csv‘ and split it into features and labels</li> 
<li> Convert data to floating point numbers </li>
<li> Drop all not useful features </li>
</ol>
<div>

<div class = "alert alert-danger" style="border-radius:10px"> **HINT:** You may want to consider
pandas library for cleaning purposes.
</div>

In [37]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler

# Read the dataset ‘churn.csv‘
data = pd.read_csv('churn.csv')

# clean data
# drop features: state, account length, area code, phone
to_drop = ['State', 'Account Length', 'Area Code', 'Phone']
clean_data = data.drop(to_drop, axis=1)

boolean_cols = ["Int'l Plan","VMail Plan", 'Churn?']
for col in boolean_cols:
    # convert categorical data into numeric data
    clean_data[col] = clean_data[col].astype('category').cat.codes

# convert to float data
clean_data = clean_data.astype(np.float)
# split it into features and labels
X, y = clean_data.drop(['Churn?'], axis=1), clean_data['Churn?']

# set y values as 1 and -1
y[y == 0] = -1

#Normalization of the data via Ponderated Normalization
for i in range(X.shape[1]):
    X.ix[:,i] = (X.ix[:,i] - X.ix[:,i].min()) / (X.ix[:,i].max() - X.ix[:,i].min());

test_size=.25
cut = int(len(X) * (1.0-test_size))
X = X.as_matrix()
X_train, X_test, y_train, y_test = X[:cut], X[cut:], y[:cut], y[cut:]

<div class = "alert alert-success" style = "border-radius:10px">**WP 2**
<ol>
<li> Try solving the problem using sklearn techniques.</li> 
</ol>
<div>

In [39]:
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.gaussian_process import GaussianProcessClassifier
from sklearn.gaussian_process.kernels import RBF
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis

#define classifiers
classifiers = [
    (KNeighborsClassifier(3), "Nearest Neighbours"),
    (SVC(kernel="linear", C=0.025), "Linear SVM"),
    (SVC(gamma=2, C=1), "RBF SVM"),
    (GaussianProcessClassifier(1.0 * RBF(1.0), warm_start=True), "Gaussian Process"),
    (DecisionTreeClassifier(max_depth=5), "Decision Tree"),
    (RandomForestClassifier(max_depth=5, n_estimators=10, max_features=1), "Random Forest"),
    (MLPClassifier(alpha=1), "Neural Network"),
    (AdaBoostClassifier(), "AdaBoost"),
    (GaussianNB(), "Naive Bayes"),
    (QuadraticDiscriminantAnalysis(), "Quadratic Discriminant Analysis")]

#Train and test classifiers                                                         
print "Scores of sklearn techniques:\n"
for csf in classifiers:
    csf[0].fit(X_train,y_train)
    print csf[1] + ": " + str(csf[0].score(X_test,y_test))

Scores of sklearn techniques:

Nearest Neighbours: 0.892086330935
Linear SVM: 0.836930455635
RBF SVM: 0.906474820144
Gaussian Process: 0.932853717026
Decision Tree: 0.932853717026
Random Forest: 0.840527577938
Neural Network: 0.83932853717
AdaBoost: 0.880095923261
Naive Bayes: 0.863309352518
Quadratic Discriminant Analysis: 0.858513189448


<div class = "alert alert-success" style = "border-radius:10px">**WP 3**
<ol>
<li> Implement your classifier.</li> 
</ol>
<div>

<div class = "alert alert-success" style = "border-radius:10px">***3.1 - Adaboost***

In [28]:
import numpy as np;
from sklearn.base import BaseEstimator, ClassifierMixin;
from sklearn.utils.validation import check_X_y, check_array, check_is_fitted;
from sklearn.utils.multiclass import unique_labels;
from sklearn.metrics import euclidean_distances;
from sklearn import linear_model;
from sklearn import metrics;

#This ADABOOST implementation considers binary classes {-1, 1}

class myAdaboostClassifier(BaseEstimator, ClassifierMixin):
    
    def __init__(self, T=5, demo_param='demo'):
        self.demo_param = demo_param;
        self.T = T; #Number of ADABOOST stumps
        self.adaboost_stumps = np.zeros((T,5)); #ADABOOST stumps(feature-best classifier,threshold,alpha,log_reg coef)
        self.stump_generator = linear_model.LogisticRegression();  #ADABOOST Stump Threshold Generator - LogReg approach : ax + b = 0; => x or th = -b / a;

    def fit(self, X, y):

        X, y = check_X_y(X, y)
        
        self.classes_ = unique_labels(y)
        self.X_ = X
        self.y_ = y
        
        W = (1.0 / len(self.y_)) * np.ones(self.y_.shape[0]);

        for t in range(self.T):

                lessMisclassifiedSample = [];
                lessClassEntropy = [];
                minErr = 1;
                bestFeat = 0;
                bestCoef = 0;
                bestAlpha = 1;
                bestThreshold = 0;

                for i in range(self.X_.shape[1]):

                        feature_i = np.transpose(np.atleast_2d(self.X_[:,i]));
                        self.stump_generator.fit(feature_i,self.y_);
                        threshold = -1 * (self.stump_generator.intercept_ / self.stump_generator.coef_);
                        
                        y_e = self.stump_generator.predict(feature_i);
                        
                        class_entropy = y_e*self.y_;
                        missclassified_samples = [1 if ce == -1 else 0 for ce in class_entropy];
                        J_err = np.dot(missclassified_samples,W);
                        alpha_f = (1.0/2) * np.log((1-J_err) / J_err);

                        if J_err < minErr:

                                minErr = J_err;
                                bestFeat = i;
                                bestCoef = self.stump_generator.coef_;
                                bestAlpha = alpha_f;
                                bestThreshold = threshold;
                                lessMisclassifiedSample = missclassified_samples;
                                lessClassEntropy = class_entropy;
                                
                
                self.adaboost_stumps[t,:] = np.array([bestFeat,bestThreshold,bestAlpha,bestCoef,minErr]);

                for i in range(self.y_.shape[0]):
                        W[i] = W[i] * np.exp(bestAlpha*(-1 * lessClassEntropy[i]));

                Wnorm = np.sum(W);
                W = W / Wnorm;
        
        
        return self

    def predict(self, X):
        
        check_is_fitted(self, ['X_', 'y_']);
        
        X = check_array(X);
        
        y_e = np.zeros(X.shape[0]);
        
        higherClass = 1;
        lowerClass = -1;
                
        for i in range(X.shape[0]):
            
            stumpsClassLinearCom = 0;
            
            x_class = 0;
            
            for t in range(self.T):
                
                if  X[i,self.adaboost_stumps[t,0]] * self.adaboost_stumps[t,3] < 0:
                    
                    if X[i,self.adaboost_stumps[t,0]] > self.adaboost_stumps[t,1]:
                        x_class = lowerClass;
                    else:
                        x_class = higherClass;
                        
                else:
                    
                    if X[i,self.adaboost_stumps[t,0]] > self.adaboost_stumps[t,1]:
                        x_class = higherClass;
                    else:
                        x_class = lowerClass;
                     
                stumpsClassLinearCom = stumpsClassLinearCom + x_class * self.adaboost_stumps[t,2];
                    
            if stumpsClassLinearCom < 0:
                    y_e[i] = lowerClass;
            else:
                    y_e[i] = higherClass;
        
        
        return y_e;
        
    def score(self, X, y):
        return sum(self.predict(X) == y) / len(y);
        
    def get_params(self, deep=True):
        # suppose this estimator has parameters "alpha" and "recursive"
        return {"alpha": self.alpha, "recursive": self.recursive}

    def set_params(self, **parameters):
        for parameter, value in parameters.items():
            self.setattr(parameter, value)
        return self

<div class = "alert alert-success" style = "border-radius:10px">***3.2 - Random Forest***

In [29]:
# include BaseEstimator class and for generating random numbers according to scikit guidelines
import sklearn

# Calculate the Gini index for a split dataset
def weighted_gini_index(groups):
    gini = 0.0
    for group in groups:
        labels = group[:, -1]
        if len(labels):
            class_values, counts = np.unique(labels, return_counts=True)
            class_prob = counts / float(len(labels))
            weight = 1.0 * len(group) / len(groups)
            gini += weight * (1.0 - np.dot(class_prob, class_prob))
    return gini

# Calculate the Entropy for a split dataset
def weighted_entropy(groups):
    entropy = 0.0
    for group in groups:
        labels = group[:, -1]
        if len(labels):
            class_values, counts = np.unique(labels, return_counts=True)
            class_prob = counts / float(len(labels))
            weight = 1.0 * len(group) / len(groups)
            entropy += -weight * np.dot(class_prob, np.log2(class_prob))
    return entropy

class MyDecisionTreeClassifier(sklearn.base.BaseEstimator):
    def __init__(self, max_features=None, max_depth=None, min_samples_leaf=1, boostrap=False, criterion='gini', random_state=None):
        self.max_features = max_features
        self.max_depth = max_depth
        self.min_samples_leaf = min_samples_leaf
        self.boostrap = boostrap
        self.criterion = criterion
        self.random_state = random_state
        
        #internal representation
        self._num_features = None
        self._score_f = None
        self._tree = None
                
    def fit(self, X, y):
        # generate random numbers according to scikit guidelines        
        self.random_state = sklearn.utils.check_random_state(self.random_state)
        
        if self.criterion == 'gini':
            self._score_f = weighted_gini_index
        elif self.criterion == 'entropy':
            self._score_f = weighted_entropy
        else:
            self._score_f = weighted_gini_index
        
        # default: all of features
        self._num_features = len(X[0]) if self.max_features is None else self.max_features
        
        dataset = np.column_stack((X, y))

        # TODO: when boostrap is false, should data be shuffled?
        if self.boostrap:
            sample = self._subsample(dataset)

        self._tree = self._build_internal(sample)
        
        return self
    
    def predict(self, X):
        return np.array([self._rec_predict(self._tree, row) for row in X])
        
    def score(self, X, y):
        return np.sum(self.predict(X) == y) / float(len(y))
    
    # internal methods
    
    # Split a dataset based on an attribute and an attribute value
    def _test_split(self, index, value, dataset):
        feature = dataset[:, index]
        return dataset[feature < value], dataset[feature >= value]

    # Select the best split point for a dataset
    def _get_split(self, dataset):
        b_index, b_value, b_score, b_groups = 999, 999, 999, None

        features = self.random_state.choice(range(len(dataset[0])-1), self._num_features, replace=False)

        for index in features:
            for row in dataset:
                groups = self._test_split(index, row[index], dataset)
                score = self._score_f(groups)
                if score < b_score:
                    b_index, b_value, b_score, b_groups = index, row[index], score, groups
        return {'index':b_index, 'value':b_value, 'groups':b_groups}

    # Create a terminal node value
    def _to_terminal(self, group):
        outcomes = [row[-1] for row in group]
        return max(set(outcomes), key=outcomes.count)

    # Create child splits for a node or make terminal
    def _split(self, node, depth):
        left, right = node['groups']
        del(node['groups'])
        # check for a no split
        if len(left) == 0 or len(right) == 0:
            node['left'] = node['right'] = self._to_terminal(np.vstack([left, right]))
            return
        # check for max depth
        if self.max_depth is not None and depth >= self.max_depth:
            node['left'], node['right'] = self._to_terminal(left), self._to_terminal(right)
            return
        # process left child
        if len(left) <= self.min_samples_leaf:
            node['left'] = self._to_terminal(left)
        else:
            node['left'] = self._get_split(left)
            self._split(node['left'], depth+1)
        # process right child
        if len(right) <= self.min_samples_leaf:
            node['right'] = self._to_terminal(right)
        else:
            node['right'] = self._get_split(right)
            self._split(node['right'], depth+1)

    # Make a prediction with a decision tree
    def _rec_predict(self, node, row):
        if row[node['index']] < node['value']:
            if isinstance(node['left'], dict):
                return self._rec_predict(node['left'], row)
            else:
                return node['left']
        else:
            if isinstance(node['right'], dict):
                return self._rec_predict(node['right'], row)
            else:
                return node['right']

    # Create a random subsample from the dataset with replacement
    def _subsample(self, dataset):
        n_sample = len(dataset)
        index = self.random_state.choice(range(n_sample), n_sample, replace=self.boostrap)
        sample = dataset[index]
        return sample 

    # Build a decision tree
    def _build_internal(self, train):
        root = self._get_split(train)
        self._split(root, 1)
        return root
    
class MyRandomForestClassifier(sklearn.base.BaseEstimator):

    def __init__(self, n_estimators=10, max_features=None, max_depth=None, min_samples_leaf=1, boostrap=True, criterion='gini', random_state=None):
        self.n_estimators = n_estimators
        self.max_features = max_features
        self.max_depth = max_depth
        self.min_samples_leaf = min_samples_leaf
        self.boostrap = boostrap
        self.criterion = criterion
        self.random_state = random_state
        
        # internal representation
        self._estimators = []
        
    def fit(self, X, y):
        # generate random numbers according to scikit guidelines
        self.random_state = sklearn.utils.check_random_state(self.random_state)        

        # default: sqrt of the total number of features
        num_features = int(np.sqrt(len(X[0]))) if self.max_features is None else self.max_features
            
        params = dict(max_depth=self.max_depth, min_samples_leaf=self.min_samples_leaf, 
                      max_features=num_features, boostrap=self.boostrap, criterion=self.criterion,
                      random_state=self.random_state) 

        self._estimators = [MyDecisionTreeClassifier(**params).fit(X, y) for t in range(self.n_estimators)]

        return self

    def predict(self, X):        
        return np.array([self._bagging_predict(sample) for sample in X]) 

    def score(self, X, y):
        return np.sum(self.predict(X) == y) / float(len(y))
    
    # Make a prediction with a list of bagged trees
    def _bagging_predict(self, sample):
        class_values, counts = np.unique([tree.predict([sample]) for tree in self._estimators], return_counts=True)
        return class_values[np.argmax(counts)]

<div class = "alert alert-success" style = "border-radius:10px">**WP 4**
<ol>
<li> Test your classifier.</li> 
</ol>
<div>

In [38]:
POS, NEG = 1, -1

conf_matrix = """
            True class
              0   1
Predicted  0 {tn}  {fn} 
class      1 {fp}  {tp}       
"""

#prints confustion matrix, accuracy, precision and recall for the given predictions
def print_stats(predictions):
    TP = np.sum( y_test[predictions == POS] == POS )
    TN = np.sum( y_test[predictions == NEG] == NEG )
    FP = np.sum( y_test[predictions == POS] == NEG )
    FN = np.sum( y_test[predictions == NEG] == POS )

    print(conf_matrix.format(tn=TN, tp=TP, fn=FN, fp=FP))
    print "Accuracy: ", 1. * (TN + TP) / (TN + TP + FN + FP)
    print "Precision: ", 1. * TP / (TP + FP)
    print "Recall: ", 1. * TP / (TP + FN)

  
#testing adaboost
print('ADABOOST')
ada = myAdaboostClassifier(5).fit(X_train,y_train)
print_stats(ada.predict(X_test))

print("\n\n")

#testing random forest
print('RANDOM FOREST')
rnf = MyRandomForestClassifier(n_estimators=10, criterion='entropy').fit(X_train, y_train)
predictions = rnf.predict(X_test)
print_stats(predictions)

ADABOOST


/home/felix/anaconda2/lib/python2.7/site-packages/ipykernel/__main__.py:93: VisibleDeprecationWarning: using a non-integer number instead of an integer will result in an error in the future
/home/felix/anaconda2/lib/python2.7/site-packages/ipykernel/__main__.py:102: VisibleDeprecationWarning: using a non-integer number instead of an integer will result in an error in the future



            True class
              0   1
Predicted  0 697  134 
class      1 1  2       

Accuracy:  0.838129496403
Precision:  0.666666666667
Recall:  0.0147058823529



RANDOM FOREST

            True class
              0   1
Predicted  0 692  51 
class      1 6  85       

Accuracy:  0.931654676259
Precision:  0.934065934066
Recall:  0.625


<div class = "alert alert-success" style = "border-radius:10px">**WP 5**
<ol>
<li> Evaluate your code comparing the results with a couple of techniques from sklearn.</li> 
</ol>
<div>

Let us now compare our results with the results of sklearn techniques observed in WP2:
<ul>
<li>Nearest Neighbours: 0.892086330935</li>
<li>Linear SVM: 0.836930455635</li>
<li>RBF SVM: 0.906474820144</li>
<li>Gaussian Process: 0.932853717026</li>
<li>Decision Tree: 0.932853717026</li>
<li>Random Forest: 0.840527577938</li>
<li>Neural Network: 0.83932853717</li>
<li>AdaBoost: 0.880095923261</li>
<li>Naive Bayes: 0.863309352518</li>
<li>Quadratic Discriminant Analysis: 0.858513189448</li>
</ul>

As we can see, the scores of the sklearn techniques vary strongly, so we will only compare our classifier with the highest scoring techniques.

TODO

The problem is an unbalanced problem and our goal is to miss as few as possible
churn clients. For each predicted churn client the company will spend resources to try to
keep him engaged.
Consider that a given client yields a benefit of 100 euros.

<div class = "alert alert-success" style = "border-radius:10px">**WP 6**
<ol>
<li> Propose a solution for different costs (in euros) for regaining a client</li> 
<li> Look for the most profitable retention campaign (How much can be invest in
retention?)</li> 

</ol>
<div>

In [43]:
#static

#financial benefit of preventing a churn
ben = 100

TP = np.sum( y_test[predictions == POS] == POS )
FP = np.sum( y_test[predictions == POS] == NEG )


#dynamic

#churn prevention rate after actions are performed
cpr = 0.2

#cost for performing certain actions on potential churners
c = 10


benefit = ben*cpr*TP
cost = c*(TP+FP)
profit = benefit - cost

print profit

790.0


<div class = "alert alert-success" style = "border-radius:10px">**Results**
<ol>
<li> Table comparing performances. This includes, for instance, error rate and standard deviation. </li> 
<li> Graphical presentation of the information in the previous table. </li> 
<li> Graphical representation of the benefit vs the cost of the retention campaign. </li>
</ol>
<div>

## V. Conclusions and Future Work
<div style="text-align:justify">
<p></p>
<p>After defining properly the project goals, having an accurate representation of the data by performing data standarization, reviewing all the scikit-learn literature pre-developed methodlogies and selecting the most fittable algorithm to perform churn-classification, we come to the conclusion of..</p>

</div>

## VI. Bibliography
<div style="text-align:justify">
    <p></p>
    <ol>
        <li>Wikipedia. (2016, December). Churn rate [Online]. Available: https://en.wikipedia.org/wiki/Churn_rate</li>
        <li>BioMedWare. (2011, November). Data Standarization [Online]. Availabe: https://www.biomedware.com/files/documentation/boundaryseer/Preparing_data/How_to_standardize_variables.htm </li>
        <li>Robert E. Schapire. A Brief Introduction to Boosting. <i>Sixteenth International Joint Conference on Artificial Intelligence</i>. New Jersey, U.S.A. 1999.</li>
        <li>Robert E. Schapire. A Brief Explaining ADABOOST. <i>Princeton  University,  Dept.  of  Computer  Science</i>. New Jersey, U.S.A. 2012.</li>
        <li>Wikipedia. (2016, December). Random Forests [Online]. Available: https://en.wikipedia.org/wiki/Random_forest</li>
        <li>Wikipedia. (2016, December). Decision Trees [Online]. Available: https://en.wikipedia.org/wiki/Decision_tree_learning</li>
        <li>Wikipedia. (2016, December). ID3 [Online]. Available: https://en.wikipedia.org/wiki/ID3_algorithm </li>
    </ol>

</div>